sycl/test/regression/hier_par.cpp

```diff
@@ -0,0 +1,105 @@
+//==- hier_par.cpp --- hierarchical parallelism regression tests -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: %clangxx -O2 -fsycl -fsycl-device-only -c -Xclang -emit-llvm -o %t.ll %s
+// RUN: cat %t.ll | FileCheck %s
+
+// This test checks for bug fix regressions related to hierarchical parallelism.
+// - bug1: private var's (cl::sycl::group argument) address shared locally
+//   the test checks that a "shadow" local variable is generated for the group
+//   argument
+//
+// This is compile-only test for now.
+//
+
+#include <CL/sycl.hpp>
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+
+using namespace cl::sycl;
+
+const int N_WG = 7;
+const int WG_SIZE = 3;
+
+const int N_ITER = 1000;
+
+class MySelector : public device_selector {
+  virtual int operator()(const device &device) const {
+    if (const char *dev_type = getenv("SYCL_DEVICE_TYPE")) {
+      if (!strcmp(dev_type, "GPU"))
+        return device.is_gpu() ? 1000 : -1;
+      if (!strcmp(dev_type, "CPU"))
+        return device.is_cpu() ? 1000 : -1;
+      if (!strcmp(dev_type, "ACC"))
+        return device.is_accelerator() ? 1000 : -1;
+      if (!strcmp(dev_type, "HOST"))
+        return device.is_host() ? 1000 : -1;
+      std::cerr << "Unsupported device type: " << dev_type << "\n";
+      return -1;
+    }
+    return 1;
+  }
+};
+
+int main() {
+  const int range_length = N_WG * WG_SIZE;
+  std::unique_ptr<int> data(new int[range_length]);
+  int *ptr = data.get();
+  std::memset(ptr, 0, range_length * sizeof(ptr[0]));
+
+  try {
+    queue myQueue(MySelector{});
+    buffer<int, 1> buf(ptr, range<1>(range_length));
+
+    myQueue.submit([&](handler &cgh) {
+      std::cout << "Running on "
+                << myQueue.get_device().get_info<cl::sycl::info::device::name>()
+                << "\n";
+      auto dev_ptr = buf.get_access<access::mode::read_write>(cgh);
+
+      cgh.parallel_for_work_group<class hpar_hw>(
+          range<1>(N_WG), range<1>(WG_SIZE), [=](group<1> g) {
+// CHECK: @[[SHADOW:[a-zA-Z0-9]+]] = internal unnamed_addr addrspace(3) global %[[GROUP_CLASS:"[^"]+"]] undef, align [[ALIGN:[0-9]+]]
+// CHECK: define {{.*}} spir_func void @{{"[^"]+"}}({{[^,]+}}, %[[GROUP_CLASS]]* byval(%[[GROUP_CLASS]]) align {{[0-9]+}} %[[GROUP_OBJ:[A-Za-z_0-9]+]]) {{.*}}!work_group_scope{{.*}} {
+// CHECK-NOT: define {{.*}} spir_func
```

Comment:
Well, today there can't be more kernels, as there is just one in the source and compiler does not generate extra, so I thought this would be reliable enough. But I'll change to make sure no `define` gets in the way - would be more reliable indeed.